sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala

```diff
@@ -683,7 +699,28 @@ abstract class HadoopFsRelation private[sql](
       }
     }
 
-    buildInternalScan(requiredColumns, filters, inputStatuses, broadcastedConf)
+    if (bucketSpec.isDefined && sqlContext.conf.bucketingEnabled()) {
+      groupBucketFiles(inputStatuses).map { groupedBucketFiles =>
+        // For each bucket id, firstly we get all files belong to this bucket, by detecting bucket
+        // id from file name. Then read these files into a RDD(use one-partition empty RDD for empty
+        // bucket), and coalesce it to one partition. Finally union all bucket RDDs to final result.
+        val perBucketRows = (0 until bucketSpec.get.numBuckets).map { bucketId =>
+          val inputStatuses = groupedBucketFiles.get(bucketId)
+          if (inputStatuses.isEmpty) {
+            sqlContext.emptyResult
+          } else {
+            buildInternalScan(
+              requiredColumns, filters, inputStatuses.get, broadcastedConf).coalesce(1)
+          }
+        }
+
+        new UnionRDD(sqlContext.sparkContext, perBucketRows)
+      }.getOrElse {
+        buildInternalScan(requiredColumns, filters, inputStatuses, broadcastedConf)
```

Comment:
Can you just fallback this way? If the metadata says it is bucketed, it needs to be bucketed.